File: code/150/solution.go
package main /*{{{*/

import (
	"fmt"
	"math"
	"reflect"
)

const (
	e4  = int64(10000)
	e5  = int64(100000)
	e6  = int64(1000000)
	e7  = int64(10000000)
	e8  = e4 * e4
	mod = int64(1e9 + 7)
)

func sqrt(a int64) int64 {
	x := math.Sqrt(float64(a))
	return int64(x)
}

func add(list ...interface{}) int64 { /*{{{*/
	if len(list) == 0 {
		return 0
	}
	ret := int64(0)
	for _, i := range list {
		buf := int64(0)
		switch reflect.TypeOf(i).String() {
		case "int":
			buf = int64(i.(int))
		case "int64":
			buf = i.(int64)
		}
		ret = (ret + buf) % mod
	}
	return ret
} /*}}}*/
func mul(list ...interface{}) int64 { /*{{{*/
	if len(list) == 0 {
		return 0
	}
	ret := int64(1)
	for _, i := range list {
		buf := int64(0)
		switch reflect.TypeOf(i).String() {
		case "int":
			buf = int64(i.(int))
		case "int64":
			buf = i.(int64)
		}
		buf %= mod
		base := ret
		ret = 0
		for buf > 0 {
			if (buf & 1) == 1 {
				ret = add(ret, base)
			}
			base = add(base, base)
			buf >>= 1
		}
	}
	return ret
} /*}}}*/
func qpow(a, b int64) int64 { /*{{{*/
	ret := int64(1)
	for b > 0 {
		if (b & 1) == 1 {
			ret = mul(ret, a)
		}
		a = mul(a, a)
		b >>= 1
	}
	return ret
} /*}}}*/
/*}}}*/

var t = int64(0)
var mod1 = int64(1 << 20)
var mod2 = int64(1 << 19)

func gen() int64 {
	t = (615949*t + 797807) % mod1
	return t - mod2
}

var sum = [1010][1010]int64{}

func main() {
	for i := 1; i <= 1000; i++ {
		for j := 1; j <= i; j++ {
			sum[i][j] = sum[i][j-1] + gen()
		}
	}
	ans := int64(2e9)
	for i := 1000; i >= 1; i-- {
		for j := 1; j <= i; j++ {
			now := int64(0)
			for l := 0; l <= 1000-i; l++ {
				now += sum[i+l][j+l] - sum[i+l][j-1]
				if now < ans {
					ans = now
				}
			}
		}
	}
	fmt.Println(ans)
}

File: code/150/solution.cpp
#include "../template/bigint.cpp"
#include "../template/china_reminder.cpp"
#include "../template/euler.cpp"
#include "../template/pell.cpp"
#include "../template/prime.cpp"
#include "../template/prime_cnt.cpp"
#include <iostream>/*{{{*/
#include <map>
#include <math.h>
#include <set>
#include <vector>
const long long e2 = 100;
const long long e3 = e2 * 10;
const long long e4 = e3 * 10;
const long long e5 = e4 * 10;
const long long e6 = e4 * e2;
const long long e7 = e6 * 10;
const long long e8 = e7 * 10;
const long long e9 = e8 * 10;
long long _mod     = e9 + 7;
using namespace std; /*}}}*/

long long t = 0;
int mod1    = 1 << 20;
int mod2    = 1 << 19;
int gen() {
    t = (615949 * t + 797807) % mod1;
    return t - mod2;
}
long long sum[1010][1010];
int main() {
    for (int i = 1; i <= 1000; i++) {
        for (int j = 1; j <= i; j++) {
            sum[i][j] = sum[i][j - 1] + gen();
        }
    }
    long long ans = 2e9;
    for (int i = 1000; i >= 1; i--) {
        for (int j = 1; j <= i; j++) {
            long long now = 0;
            for (int l = 0; l <= 1000 - i; l++) {
                now += sum[i + l][j + l] - sum[i + l][j - 1];
                ans = min(ans, now);
            }
        }
    }
    cout << ans << endl;
    return 0;
}
